# <center>SQL Analysis On B2B Sales Data</center>

## Set Up

Import necessary libraries

In [1]:
# importing pandas as pd just in case it is needed
import pandas as pd
from sqlalchemy import create_engine

Load SQL extenstion

In [2]:
%load_ext sql

Establish conncection to MySQL Database

In [3]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

## Part 1: Exploratory Analysis

a). Quite a few accounts have NULL in the PartnerId column so it is worth it to take a closer look to see exactly how many null values are there in the PartnerId column and what these NULL values mean.

In [10]:
%%sql
SELECT COUNT(*)
FROM Accounts
WHERE PartnerId IS NULL;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[(3035,)]

In [11]:
%%sql
SELECT 
    AccountName,
    PartnerId
FROM Accounts
WHERE PartnerId IS NULL;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
3035 rows affected.


[('Blue Origin', None),
 ('Agilent Technologies, Inc', None),
 (' Project Systems Corporation', None),
 ('Alibaba Group', None),
 ('Canon Business Professional Services (OEM)', None),
 ('CloudDOCX (OEM)', None),
 ('Crunch Technologies Corporation', None),
 ('Curiato, Inc. (OEM)', None),
 ('eLeadCRM (OEM)', None),
 ('Elixir Technologies Corporation', None),
 ('EZOPS Inc (OEM)', None),
 ('Flatirons Solutions Inc (OEM)', None),
 ('IPEC Project Systems Sdn Bhd', None),
 ('Iron Mountain, Inc', None),
 ('Kronos Incorporated', None),
 ('Link Technology Group', None),
 ('MoxiWorks', None),
 ('Mutual Concept Computer Group Inc. (OEM)', None),
 ('National Association of Insurance Commissioners', None),
 ('National Sports Academy', None),
 ('New Machine Computer Group Inc.', None),
 ('Omnicon Group (OEM)', None),
 ('Orbis Technologies, Inc.', None),
 ('RioSoft (OEM)', None),
 ('Sapiens Americas', None),
 ('Thomson Reuters', None),
 ('Toppan Merrill', None),
 ('West Coast Electric Corporation', None),
 ('Xerox DocuShare (OEM)', None),
 ('Yokogawa Electric Corporation', None),
 ('Accurate Background', None),
 ('Lexipol LLC', None),
 ('Los Angeles Unified School District', None),
 ('Michaels Stores, Inc.', None),
 ('NextSphere Technologies', None),
 ('SAE International', None),
 ('Social Finance, Inc.', None),
 ('SOUTHWEST BUSINESS CORPORATION', None),
 ('UCLA', None),
 ('University of California, Agriculture and Natural Resources', None),
 ('4DX', None),
 ('Accenture China', None),
 ('ACCENTURE PTE LTD', None),
 ('AIA Hong Kong', None),
 ('Allianz SE Singapore Branch', None),
 ('Aqsacom', None),
 ('Beijing Shouhuan Cultural Tourism Investment Co., Ltd.', None),
 ('Bureau Veritas HK Ltd.', None),
 ('Central Bank', None),
 ('Cognizant Technology Solutions Hong Kong Ltd', None),
 ('Department of Parliamentary Services', None),
 ('Eclipx Group Limited', None),
 ('Enterprise Singapore', None),
 ('GE Digital China', None),
 ('GENIX VENTURES PTY LTD', None),
 ('i3esolutions, inc', None),
 ('Icare NSW', None),
 ('James Cook University Singapore', None),
 ('Li & Fung (Trading) Limited', None),
 ('LINK MARKET SERVICES GROUP PTY LIMITED', None),
 ('MINTER ELLISON LAWYERS', None),
 ('NATIONAL AUSTRALIA BANK LIMITED', None),
 ('NCSI', None),
 ('News Pty Limited', None),
 ('NSW DEPARTMENT OF POLICE AND JUSTICE', None),
 ('Prudential Investment Company of Australia', None),
 ('SIEBRENET Inc.', None),
 ('Standard Chartered Bank', None),
 ('TABCORP HOLDINGS LIMITED', None),
 ('TELSTRA HOLDINGS PTY LTD', None),
 ('UFINITY GROUP LIMITED', None),
 ('Vietnam Electricity', None),
 ('Welab Holdings', None),
 ('Bice Vida Compañía De Seguros', None),
 ('Colmena Seguros', None),
 ('Pentafon', None),
 ('Procargo Port', None),
 ('SOGE Bank', None),
 ('St. Lucia - Government of Information Technology', None),
 ('ZURICH SANTANDER - Uruguay', None),
 ('Boeing DMS', None),
 ('California DOJ', None),
 ('CoreLogic Inc, Solutions, LLC', None),
 ('Google Inc', None),
 ('Intel Corporation', None),
 ('Oregon Department of Human Services', None),
 ('Robert Half International Inc', None),
 ('TOYOTA MOTOR CREDIT CORPORATION', None),
 ('Wells Fargo & Company AZ', None),
 ('Wells Fargo Bank, N.A.', None),
 ('Boost Insurance', None),
 ('Quality Steel Corporation', None),
 ('ScienceLogic', None),
 ('Spruce Technology', None),
 ('SoftBank', None),
 ('Citigroup', None),
 ('Facebook', None),
 ('Intel', None),
 ('OurCrowd', None),
 ('Pindrop', None),
 ('Stripe', None),
 ('The Honest Company', None),
 ('Jumio', None),
 ('Airwallex', None),
 ('Groww', None),
 ('Redis Labs', None),
 ('Hopper', None),
 ('Ironclad', None),
 ('Cyble', None),
 ('DocuSign', None),
 ('Silver Lake', None),
 ('Box', None),
 ('CrowdStrike', None),
 ('Outreach', None),
 ('Unacademy', None),
 ('Verkada', None),
 ('Seedrs', None),
 ('Skydio', None),
 ('GupShup', None),
 ('Cowbell Cyber', None),
 ('SecurityScorecard', None),
 ('Draper Esprit', None),
 ('Trax', None),
 ('Dropbox', None),
 ('Airtable', None),
 ('Bitwise I

In [12]:
%%sql
SELECT *
FROM Accounts;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
4112 rows affected.


[(1, 1, 'Ascend Learning', 5, 'Closed Won', 'Closed', 10625, 16, datetime.datetime(2019, 9, 20, 0, 0), 8, 'Existing Customer - Upsell', datetime.datetime(2019, 9, 17, 0, 0), 1),
 (2, 1, 'NetPower', 98, '(1) Gain Access', 'Pipeline', 83250, 12, datetime.datetime(2019, 9, 13, 0, 0), 25, 'New Customer', datetime.datetime(2019, 10, 31, 0, 0), 2),
 (3, 1, 'Blue Origin', 134, '(4) Propose Solution', 'Best Case', 153136, 12, datetime.datetime(2019, 7, 31, 0, 0), None, 'New Customer', datetime.datetime(2019, 11, 22, 0, 0), 3),
 (4, 1, 'Agilent Technologies, Inc', 119, '(3) Develop Solution', 'Pipeline', 30000, 12, datetime.datetime(2019, 11, 15, 0, 0), None, 'Existing Customer - New Project', datetime.datetime(2019, 11, 15, 0, 0), 4),
 (5, 2, 'AyP Mexico', 19, '(1) Gain Access', 'Pipeline', 15300, 12, datetime.datetime(2019, 11, 30, 0, 0), 14, 'New Customer', datetime.datetime(2019, 11, 30, 0, 0), 5),
 (6, 3, 'Aboitiz Power Corporation', 406, '(5) Negotiate & Close', 'Pipeline', 90478, 60, datetime.datetime(2019, 4, 29, 0, 0), 11, 'New Customer', datetime.datetime(2019, 10, 24, 0, 0), 6),
 (7, 3, 'Banco de Oro', 288, '(5) Negotiate & Close', 'Commit', 29644, 24, datetime.datetime(2019, 6, 14, 0, 0), 1, 'New Customer', datetime.datetime(2019, 10, 17, 0, 0), 6),
 (8, 4, 'Falabella', 83, '(1) Gain Access', 'Pipeline', 70125, 12, datetime.datetime(2019, 12, 13, 0, 0), 2, 'New Customer', datetime.datetime(2020, 2, 7, 0, 0), 5),
 (9, 1, 'State of Connecticut Department of Children & Families', 28, '(1) Gain Access', 'Pipeline', 30000, 12, datetime.datetime(2020, 2, 21, 0, 0), 3, 'New Customer', datetime.datetime(2020, 2, 21, 0, 0), 1),
 (10, 1, 'PUBLIC EMPLOYEES RETIREMENT SYSTEM, CALIFORNIA', 427, '(4) Propose Solution', 'Pipeline', 34200, 12, datetime.datetime(2019, 9, 26, 0, 0), 4, 'New Customer', datetime.datetime(2019, 10, 28, 0, 0), 3),
 (11, 5, 'Teraxion', 40, '(1) Gain Access', 'Pipeline', 46750, 36, datetime.datetime(2019, 8, 31, 0, 0), 5, 'New Customer', datetime.datetime(2020, 2, 28, 0, 0), 1),
 (12, 1, 'Tucson Electric Power Company Inc', 76, '(3) Develop Solution', 'Pipeline', 200685, 36, datetime.datetime(2019, 8, 31, 0, 0), 6, 'New Customer', datetime.datetime(2019, 12, 16, 0, 0), 3),
 (13, 6, 'Rothbury Group Limited', 631, '(1) Gain Access', 'Pipeline', 29967, 36, datetime.datetime(2018, 4, 18, 0, 0), 7, 'New Customer', datetime.datetime(2020, 1, 20, 0, 0), 6),
 (14, 1, 'Crowe', 62, '(1) Gain Access', 'Pipeline', 42500, 36, datetime.datetime(2019, 8, 31, 0, 0), 8, 'New Customer', datetime.datetime(2019, 12, 27, 0, 0), 1),
 (15, 7, 'Tan Tock Seng Hospital', 459, '(1) Gain Access', 'Pipeline', 22500, 36, datetime.datetime(2018, 11, 20, 0, 0), 9, 'New Customer', datetime.datetime(2020, 1, 8, 0, 0), 6),
 (16, 8, 'America Indochina Management', 67, '(1) Gain Access', 'Pipeline', 25000, 36, datetime.datetime(2019, 8, 31, 0, 0), 10, 'New Customer', datetime.datetime(2019, 12, 11, 0, 0), 6),
 (17, 3, 'Department of Information and Communications Technology', 252, '(1) Gain Access', 'Pipeline', 170960, 36, datetime.datetime(2019, 2, 28, 0, 0), 11, 'New Customer', datetime.datetime(2020, 2, 19, 0, 0), 6),
 (18, 9, 'Industrial and Commercial Bank of China (Thai) PCL', 459, '(1) Gain Access', 'Pipeline', 36932, 36, datetime.datetime(2018, 12, 11, 0, 0), 11, 'New Customer', datetime.datetime(2019, 12, 26, 0, 0), 6),
 (19, 3, 'Integrated Microelectronics, Inc.', 98, '(1) Gain Access', 'Pipeline', 14400, 36, datetime.datetime(2020, 2, 20, 0, 0), 11, 'New Customer', datetime.datetime(2020, 2, 12, 0, 0), 6),
 (20, 10, 'MAMPU: Malaysia Administrative Modernisation And Management Planning Unit Prime Minister?s Department', 75, '(3) Develop Solution', 'Pipeline', 60015, 12, datetime.datetime(2020, 1, 5, 0, 0), 11, 'Existing Customer - Upsell', datetime.datetime(2020, 1, 5, 0, 0), 6),
 (21, 10, 'Multimedia University', 242, '(3) Develop Solution', 'Pipeline', 51500, 36, datetime.datetime(2019, 2, 28, 0, 0), 11, 'New Customer', datetime.datetime(

   Through the above queries we were able to find out that a total of 3305 records do not have a PartnerId. To further the exploration, I also generated a query to list all the accounts that do not have a ParnterId and then queried the entire table just to get a better idea of how this data is laid out. Based on my exploration, customer accounts that have been generated in-house are not tied to a partner which is represented by the NULL value. On the other hand, customer accounts that are coming through partner channels are tied to a PartnerId which links into the partner the Partners table where we can see the name of the partner tied to a specific account. In the future it may be better to explicitly document that there is no partner attached to an account. As the NULL value is a little ambiguous and it could be hard to tell if a customer account does not have an account or the record was entered incorrectly.
    
   This information can be useful as we can now compare how many of the company's sales and potential sales were/are being done through a partner vs their own resources. We can also compare the contract price to see if the company makes a larger sale through its partners or on its own. Analysis can also be done to see who the top selling partners are, and what regions working with a partner has been the most affective.

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

b). Another area worth exploring is the contract price. In order to get an idea of how much the company sells its products and services for and to look for any outliers (which could artificially raise or drop the average contract price) we can run standard descriptive measures     

In [17]:
%%sql
SELECT
    COUNT(ContractPrice) AS 'N',
    SUM(ContractPrice) AS 'sum',
    AVG(ContractPrice) AS 'mean',
    STDDEV(ContractPrice) AS 'std. dev.',
    VAR_POP(ContractPrice) AS 'varience',
    MIN(ContractPrice) AS 'min',
    MAX(ContractPrice) AS 'max'
FROM Accounts;
    

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[(4112, Decimal('790360865'), Decimal('192208.3816'), 116420.47704666048, 13553727475.772001, 1724, 399713)]

Through the above descriptive statistics of the contract price, we find:

* The total number of rows in the accounts table is 4,112
* Total dollar amount of contracts both in the pipe and closed is 790,360,865
* Average price of a contract from this company is 192,208.38
* A high standard deviation of 116,420.48 and an extremely high variance of 13,553,727,475.77
* The smallest contract was priced at 1,724
* Largest contract was priced at 399,713

The large standard deviation indicates that the contract prices are spread across a large range of values and do not hover around the mean. In addition, the variance not only confirms the wide range of values from the mean, but also shows us that the contract prices in relation to each other are very different. Based on the variance and standard deviation we can tell that there is little to no correlation between the contract prices. Although these values make sense for this specific dataset for this project since the data was randomly generated, it does not accurately represent what one would find in a live sales dataset professionally. Companies usually have set pricing based on what product or service they are buying. Therefore, we would expect larger similarity between the contract prices and the standard deviation and variance to be way smaller.

Conducting descriptive statistics is helpful not just for sales analyst, but any analyst that is working with a new dataset and is trying to get a better understanding of it. Descriptive statistics provide a great baseline for analyst to build their understanding and analysis on.   

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

c). Now that we have a better understanding of what the NULL values mean for PartnerId and have a better picture of the price data points, let’s explore what countries the company sells to.

In [20]:
%%sql
SELECT
    CountryName,
    RegionName
FROM Countries c
JOIN CountryRegions cr
    ON c.CountryId = cr.CountryId
JOIN Regions r
    ON cr.RegionID = r.RegionID;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
70 rows affected.


[('United States', 'AMER'),
 ('Canada', 'AMER'),
 ('Mexico', 'LATAM'),
 ('Chile', 'LATAM'),
 ('Peru', 'LATAM'),
 ('Colombia', 'LATAM'),
 ('Dominica', 'LATAM'),
 ('Costa Rica', 'LATAM'),
 ('Guatemala', 'LATAM'),
 ('Brazil', 'LATAM'),
 ('Haiti', 'LATAM'),
 ('Saint Lucia', 'LATAM'),
 ('Uruguay', 'LATAM'),
 ('Cayman Islands', 'LATAM'),
 ('Argentina', 'LATAM'),
 ('Philippines', 'APAC'),
 ('New Zealand', 'APAC'),
 ('Singapore', 'APAC'),
 ('Vietnam', 'APAC'),
 ('Thailand', 'APAC'),
 ('Malaysia', 'APAC'),
 ('Fiji', 'APAC'),
 ('India', 'APAC'),
 ('China', 'APAC'),
 ('Australia', 'APAC'),
 ('Hong Kong', 'APAC'),
 ('South Korea', 'APAC'),
 ('Taiwan', 'APAC'),
 ('Indonesia', 'APAC'),
 ('Poland', 'APAC'),
 ('Lithuania', 'APAC'),
 ('Bermuda', 'APAC'),
 ('Romania', 'APAC'),
 ('Croatia', 'APAC'),
 ('Belgium', 'APAC'),
 ('Latvia', 'APAC'),
 ('Pakistan', 'APAC'),
 ('Greece', 'APAC'),
 ('Portugal', 'APAC'),
 ('Bangladesh', 'APAC'),
 ('Hungary', 'APAC'),
 ('Iceland', 'APAC'),
 ('Jordan', 'APAC'),
 ('Saudi Arabia', 'APAC'),
 ('Lebanon', 'APAC'),
 ('Japan', 'JPN'),
 ('United Kingdom', 'EMEA'),
 ('Sweden', 'EMEA'),
 ('Finland', 'EMEA'),
 ('Israel', 'EMEA'),
 ('Germany', 'EMEA'),
 ('Malta', 'EMEA'),
 ('The Netherlands', 'EMEA'),
 ('Denmark', 'EMEA'),
 ('France', 'EMEA'),
 ('Switzerland', 'EMEA'),
 ('Estonia', 'EMEA'),
 ('Azerbaijan', 'EMEA'),
 ('Ireland', 'EMEA'),
 ('Spain', 'EMEA'),
 ('Italy', 'EMEA'),
 ('Ukraine', 'EMEA'),
 ('Norway', 'EMEA'),
 ('Austria', 'EMEA'),
 ('Turkey', 'EMEA'),
 ('Bahrain', 'EMEA'),
 ('United Arab Emirates', 'EMEA'),
 ('Russian Federation', 'EMEA'),
 ('Czech Republic', 'EMEA'),
 ('Slovenia', 'EMEA')]

In [22]:
%%sql
SELECT
    RegionName,
    COUNT(CountryId) AS count_countries
FROM Regions r
JOIN CountryRegions cr
    ON r.RegionId = cr.RegionId
GROUP BY RegionName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
5 rows affected.


[('AMER', 2), ('LATAM', 13), ('APAC', 30), ('JPN', 1), ('EMEA', 24)]

Based on the above queries we now know that the company sells its products and/or services in 70 different countries located in different parts of the global. The second SQL query illustrates what regions each of the countries reside in. According to the output we see that the company has closed and potential sales in:

* 2 countries in North America (AMEA)
* 13 countries in South/Latin America (LATAM)
* 30 countries in Asia and the Asia Pacific (APAC)
* 24 countries in Europe (EMEA)
* Japan (JPN which has recently been moved to its own region)

Knowing this information can be helpful as we can do a regional sales analysis to see where the company is performing well versus not so well. In addition, we can also break it down even further to examine each individual country within a given region to pinpoint exactly where the company is or is not selling. This can ultimately lead to improving marketing efforts in the underperforming area, team up with a local partner, change up the product or service offered in the country or region, or the company can decide that it is not worth trying to sell in that location anymore.

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

d). Next I will run some Max() and Min() on the ContractStartDate, CloseDate, and ContractLenghtMonths column to get an idea of the date range for this dataset. This will help determine how old these accounts are and provide insight on how long the typical contract term is for a customer.

In [9]:
%%sql
SELECT
    MAX(LEFT(ContractStartDate,4)) AS max_year_contract_start,
    MIN(LEFT(ContractStartDate,4)) AS min_year_contract_start,
    MAX(LEFT(CloseDate, 4)) AS max_year_contract_close,
    MIN(LEFT(CloseDate, 4)) AS min_year_contract_close
FROM Accounts;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[('2020', '2016', '2020', '2019')]

In [13]:
%%sql
SELECT
    MAX(ContractLengthMonths) AS max_contract_length_months,
    MIN(ContractLengthMonths) AS min_contract_length_months,
    AVG(ContractLengthMonths) AS avg_contract_length_months
FROM Accounts;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
1 rows affected.


[(60, 5, Decimal('23.8176'))]

After looking at the above SQL queries we find that the expected and/or actual contract start dates range anywhere from 2016 to 2020. However, none of these contracts were closed before 2019 as seen through the min_year_contract_close column in the first query. This provides an inconsistency within the data especially for the contracts with an expected start date ranging from 2016-2018. Since none of these contracts were closed before 2019, that would render any of the start dates before 2019 as invalid since the sale was not complete. Now the contract start dates could have been the estimated start dates, but then the deal failed to close as projected.

As for the query done for the contract length in months, we find out that the largest contract the company has or is expected to have is 60 months (roughly 5 years) while the smallest contract they have or are expected to have is only 5 months. We also find that the average sales contract at this company tends to last 23.8 months or 1.98 years. With this information we see that the company tends to prefer short-term contracts with their longest one lasting roughly 5 years.

Typically, when doing analysis, dates will play a very important role. In the context of sales data, you could use dates to compare quarters both from a realized sales perspective as well as from a pipeline projection perspective. You could also drill down even more by looking at your sales monthly to pinpoint where the sales peaks and troughs are. The sales data can also be looked at from the fiscal calendar or from the yearly calendar point of view as well.

Since the start and closed dates for this dataset are not consistent, we will just have to assume that the ContractStartDate column represents the expected start date of the contract and the same for the ClosedDate. In the future it may be helpful to beak this information up into two different columns. For example, one column would represent the expected start and closed date while the other would represent the actual start and closed date.

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

e). The final part of exploration will be focused on examining the data that surrounds the company's sales reps.

In [14]:
%%sql
SELECT *
FROM SalesReps;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
7 rows affected.


[(1, 'Robert', 'Smith'),
 (2, 'Jose', 'Gomez'),
 (3, 'Rachel', 'King'),
 (4, 'Bobby', 'Thompson'),
 (5, 'Michelle', 'Clarence'),
 (6, 'Katie', 'Barmen'),
 (7, 'Tony', 'Stamp')]

In [17]:
%%sql
SELECT
    EmployeeFirstName,
    EmployeeLastName,
    COUNT(a.EmployeeId) AS number_of_contracts
FROM SalesReps sr
JOIN Accounts a
    ON sr.EmployeeId = a.EmployeeId
GROUP BY EmployeeFirstName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
7 rows affected.


[('Robert', 'Smith', 691),
 ('Jose', 'Gomez', 658),
 ('Rachel', 'King', 654),
 ('Bobby', 'Thompson', 700),
 ('Michelle', 'Clarence', 703),
 ('Katie', 'Barmen', 705),
 ('Tony', 'Stamp', 1)]

In [32]:
%%sql
SELECT
    DISTINCT EmployeeFirstName,
    EmployeeLastName,
    RegionName,
    COUNT(r.RegionId) OVER (
        PARTITION BY RegionName, a.EmployeeId
    ) AS number_accounts_in_region
FROM SalesReps sr
JOIN Accounts a
    ON sr.EmployeeId = a.EmployeeId
JOIN Countries c
    ON a.CountryId = c.CountryId
JOIN CountryRegions cr
    ON c.CountryId = cr.CountryId
JOIN Regions r
    ON cr.RegionId = r.RegionId
ORDER BY EmployeeFirstName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
31 rows affected.


[('Bobby', 'Thompson', 'LATAM', 9),
 ('Bobby', 'Thompson', 'AMER', 472),
 ('Bobby', 'Thompson', 'JPN', 9),
 ('Bobby', 'Thompson', 'APAC', 100),
 ('Bobby', 'Thompson', 'EMEA', 110),
 ('Jose', 'Gomez', 'AMER', 459),
 ('Jose', 'Gomez', 'LATAM', 7),
 ('Jose', 'Gomez', 'APAC', 88),
 ('Jose', 'Gomez', 'JPN', 9),
 ('Jose', 'Gomez', 'EMEA', 95),
 ('Katie', 'Barmen', 'LATAM', 9),
 ('Katie', 'Barmen', 'JPN', 5),
 ('Katie', 'Barmen', 'EMEA', 104),
 ('Katie', 'Barmen', 'APAC', 130),
 ('Katie', 'Barmen', 'AMER', 457),
 ('Michelle', 'Clarence', 'EMEA', 115),
 ('Michelle', 'Clarence', 'APAC', 79),
 ('Michelle', 'Clarence', 'LATAM', 30),
 ('Michelle', 'Clarence', 'JPN', 4),
 ('Michelle', 'Clarence', 'AMER', 475),
 ('Rachel', 'King', 'EMEA', 95),
 ('Rachel', 'King', 'APAC', 71),
 ('Rachel', 'King', 'JPN', 3),
 ('Rachel', 'King', 'AMER', 477),
 ('Rachel', 'King', 'LATAM', 8),
 ('Robert', 'Smith', 'APAC', 82),
 ('Robert', 'Smith', 'AMER', 474),
 ('Robert', 'Smith', 'LATAM', 8),
 ('Robert', 'Smith', 'EMEA', 120),
 ('Robert', 'Smith', 'JPN', 7),
 ('Tony', 'Stamp', 'APAC', 1)]

For the first query I did a select * statement on the sales reps table just to see what kind of data we had strictly surrounding the sales reps. After running this query, we find that there is not a whole lot of information. The company only has 7 sales reps, and the only information we have surrounding these reps include an EmployeeId and their first and last name. In the furture it could be helpful to have more information such as their hire date, work email, and/or a work phone number.

Now that I know what data we have surrounding the company's sales reps, I wanted to see how many accounts each rep has tied to them. After running the second query we find that:

* Robert Smith has 691 contracts
* Jose Gomez has 658 contracts
* Rachel King has 654 contracts
* Bobby Thompson has 700 contracts
* Michelle Clarence has 703 contracts
* Katie Barmen has 705 contracts
* Tony Stamp has 1 contract

It is important to note that the count includes both open and closed contracts, comparing the number of open versus closed contracts per rep is something I plan on looking at in part 2 of my analysis. With this information we can get a basic idea of which sales rep have to most customer accounts. It is interesting that Tony Stamp only has 1 contract, it could be because the rep is an under performer, or it this rep may just be a new employee and has not had as long of a chance to sell like his colleagues have (therefore a hired date would have been helpful).

Finally, I wanted to break up the number of contracts per rep by region to see which region each rep was most active in. For all of the reps the North American (AMER) region seems to be where they are establishing most of their sales contracts with Europe (EMEA) coming in at a very distant second. Square which is the company that posted the job description I am basing this analysis off, is a U.S. based company. So, it would make sense that most sales for this dataset would be skewed towards AMER. In addition, it would also appear that all reps have a global presence. Normally sales reps would be assigned to sell in a specific region, however, to keep the data unbiased during the data collection phase, I randomly assigned reps to the companies I pulled from the CrunchBase API.

Overall, keeping track of sales rep performance is important to track the team's progress and success by looking at how much each representative has sold (which is what I plan to do in part two of my analysis). In addition, this information can also help the compensation team as SDR’s, and AE's make the bulk of their compensation through variable. Collecting this information can help them document and better understand the variable size that each employee should get. In addition, it is also a good way to keep track of high performers and reward them for their hard work by giving special rewards like getting invited to a "Club" event.

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

## Part 2: Answering Business Questions with Data

##### Question 1:
For the year 2020, where we able to achieve our sales forecasts for each region? If not, how far under the forecast where we?

##### Business Justification: 
Comparing your forecasted sales with your realized sales provides the sales team with feedback on how their performance was for the year. If the team was able to hit or exceed their forecasted sales number, then the sales team can re-adjust next year's forecast to be higher. On the flip side, if the team did not hit their forecasted numbers, they can also lower their forecasted numbers to offer a more attainable goal. The unclosed business (if it is not closed lost) could also be carried over to next year's forecasts thus improving the chances that the sales team will achieve their goals. This query mainly works to provide the sales executives insight into what their sales team is capable of and that they are hitting their revenue targets

##### SQL Features Used:
* CTE
* INNER JOIN
* GROUP BY

In [6]:
%%sql
WITH sales_forecast_cte AS (
    SELECT
        RegionName,
        SUM(ContractPrice) AS forecast_sales
    FROM Accounts a
    JOIN CountryRegions cr
        ON a.CountryId = cr.CountryId
    JOIN Regions r
        ON cr.RegionId = r.RegionID
    WHERE ForecastCategory IN ('Pipeline', 'Best Case', 'Commit')
        AND ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:23:59'
    GROUP BY RegionName
), 
realized_sales_cte AS (
    SELECT
        RegionName,
        SUM(ContractPrice) AS realized_sales
 FROM Accounts a
    JOIN CountryRegions cr
        ON a.CountryId = cr.CountryId
    JOIN Regions r
        ON cr.RegionId = r.RegionID
    WHERE SalesStage = 'Closed Won'
        AND ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:23:59'
    GROUP BY RegionName
)
SELECT
    sf.RegionName,
    forecast_sales,
    realized_sales,
    realized_sales > forecast_sales AS over_forecast,
    forecast_sales - realized_sales AS amount_under
FROM sales_forecast_cte sf
JOIN realized_sales_cte rs
    ON sf.RegionName = rs.RegionName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
5 rows affected.


[('AMER', Decimal('113688411'), Decimal('16034600'), 0, Decimal('97653811')),
 ('APAC', Decimal('18434031'), Decimal('2574884'), 0, Decimal('15859147')),
 ('EMEA', Decimal('26633185'), Decimal('2471762'), 0, Decimal('24161423')),
 ('LATAM', Decimal('2102126'), Decimal('347328'), 0, Decimal('1754798')),
 ('JPN', Decimal('981569'), Decimal('191793'), 0, Decimal('789776'))]

Through these results, we find that for the calendar year 2020 the sales team was able to hit their forecasted sales numbers for each of the regions. If I was a sales leader, the first thing I would do is debrief with each of my sales reps to get an idea of what next year's forecast looks like and how many of the deals that were not closed do they expect to close in the following year. Then I would readjust next year’s forecasts to be smaller and more realistic to the amount of revenue the company can generate. Based on the amount_under column it is obvious that the sales team greatly overestimated how much they could sell in a year. All the regions except for Japan under sold their forecasts by millions of dollars. Japan was only under the forecast by 789,776, but that is still an extreme misestimation by the sales team considering that JPN only represents Japan. I would suggest that the sales team re-adjust their forecast to be more reflective of how much sales the company can currently generate. Setting ridiculously high forecasts and sales targets is a good way to set the company up for failure, and destroy team moral. 

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

##### Question 2:
Let’s drill down and look at which individual regions, partners, and sales reps performed the best in terms of sales revenue generated. In addition, how did each of our partners stack up against the dollar amount of sales closed in-house without any partner support?

##### Business Justification:
By breaking down which regions are doing well by realized sales dollars, the company get a clear picture of what regions are troublesome for making sales. For example, the company can look at the lower ranked regions and do an analysis of why they are struggling to sell in those regions. It could be a culture miss match with the product or marketing strategy, high local competition, or maybe the product and/or service is just not needed in those locations.

It is also beneficial to look at partner performance because a bad partner could prolong or even kill a deal. The company is also paying these partners a percentage based off the size of the deal, so there really is no need to keep on dead weight and waste costs. In addition, comparing the dollar amount of sales brought in without a partner versus with a partner is a good way to ensure that most deals are coming through the company itself and not outside partners. More revenue can be generated through in-house sales as there are no extra costs associated with them.

Finally, it is also a good idea to keep track of your sales rep’s performance not only to generate healthy competition, but also to ensure that the employees are being rewarded appropriately for work they have put in to close deals. It can be a big morale booster when your top sales reps get rewarded with bonuses and invited to company "club" events. In addition, having these numbers handy can help sort out how much commission an employee should be or was paid for a given year.

##### SQL Features Used:
* Window Function - DENSE_RANK() & COUNT()
* LEFT & INNER JOIN
* GROUP BY
* Subquery
* VIEW

In [7]:
%%sql
SELECT
    RegionName,
    SUM(ContractPrice) AS realized_sales,
    DENSE_RANK() OVER (
        ORDER BY SUM(ContractPrice) DESC
    ) AS region_rank_by_sales
FROM Accounts a
JOIN CountryRegions cr
    ON a.CountryId = cr.CountryId
JOIN Regions r
    ON cr.RegionId = r.RegionId
WHERE SalesStage = 'Closed Won'
    AND ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59'
GROUP BY RegionName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
5 rows affected.


[('AMER', Decimal('16034600'), 1),
 ('APAC', Decimal('2574884'), 2),
 ('EMEA', Decimal('2471762'), 3),
 ('LATAM', Decimal('347328'), 4),
 ('JPN', Decimal('191793'), 5)]

In [11]:
%%sql
SELECT
    PartnerName,
    SUM(ContractPrice) AS realized_sales,
    (
        SELECT SUM(ContractPrice)
        FROM Accounts a
        LEFT JOIN Partners p
            ON a.PartnerId = p.PartnerId
        WHERE PartnerName IS NULL
            AND ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59'
            AND SalesStage = 'Closed Won'
    ) - SUM(ContractPrice) AS amount_below_nonpartner_sales,
    RANK() OVER (
        ORDER BY SUM(ContractPrice) DESC
    ) AS partner_rank_by_sales
FROM Accounts a
LEFT JOIN Partners p
    ON a.PartnerId = p.PartnerId
WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59'
    AND SalesStage = 'Closed Won'
GROUP BY PartnerName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
20 rows affected.


[(None, Decimal('15787910'), Decimal('0'), 1),
 ('Tech Mahindra Limited India', Decimal('608584'), Decimal('15179326'), 2),
 ('VietSoftware international Inc.', Decimal('492892'), Decimal('15295018'), 3),
 ('Impulse IT Ltda.', Decimal('458895'), Decimal('15329015'), 4),
 ('Ricksoft Co., Ltd', Decimal('384503'), Decimal('15403407'), 5),
 ('Crest Infosolutions SDN BHD', Decimal('369392'), Decimal('15418518'), 6),
 ('Carahsoft Technology Corporation', Decimal('350965'), Decimal('15436945'), 7),
 ('Cognizant Technology Solutions Corporation', Decimal('349577'), Decimal('15438333'), 8),
 ('Infomega SAS', Decimal('337593'), Decimal('15450317'), 9),
 ('CROZ d.o.o. (SI)', Decimal('331515'), Decimal('15456395'), 10),
 ('ClearCadence, LLC', Decimal('321652'), Decimal('15466258'), 11),
 ('Tata Consultancy Services Limited - India', Decimal('303153'), Decimal('15484757'), 12),
 ('Technology Services Group', Decimal('261575'), Decimal('15526335'), 13),
 ('iik Technologies - Desarrollo de Soluciones de Software SA de CV', Decimal('258719'), Decimal('15529191'), 14),
 ('APIUX Technologia SPA.', Decimal('243048'), Decimal('15544862'), 15),
 ('Armedia LLC', Decimal('239108'), Decimal('15548802'), 16),
 ('Crest Business Solutions Pte Ltd', Decimal('196697'), Decimal('15591213'), 17),
 ('Amihan\xa0Global\xa0Strategies\xa0Phils\xa0Inc', Decimal('122874'), Decimal('15665036'), 18),
 ('Global Quark S.A. de C.V.', Decimal('113000'), Decimal('15674910'), 19),
 ('Contezza BV', Decimal('88715'), Decimal('15699195'), 20)]

In [10]:
%%sql
SELECT
    PartnerName,
    SUM(ContractPrice) AS realized_sales,
    RANK() OVER (
        ORDER BY SUM(ContractPrice) DESC
    ) AS partner_rank_by_sales
FROM Accounts a
JOIN Partners p
    ON a.PartnerId = p.PartnerId
WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59'
    AND SalesStage = 'Closed Won'
GROUP BY PartnerName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
19 rows affected.


[('Tech Mahindra Limited India', Decimal('608584'), 1),
 ('VietSoftware international Inc.', Decimal('492892'), 2),
 ('Impulse IT Ltda.', Decimal('458895'), 3),
 ('Ricksoft Co., Ltd', Decimal('384503'), 4),
 ('Crest Infosolutions SDN BHD', Decimal('369392'), 5),
 ('Carahsoft Technology Corporation', Decimal('350965'), 6),
 ('Cognizant Technology Solutions Corporation', Decimal('349577'), 7),
 ('Infomega SAS', Decimal('337593'), 8),
 ('CROZ d.o.o. (SI)', Decimal('331515'), 9),
 ('ClearCadence, LLC', Decimal('321652'), 10),
 ('Tata Consultancy Services Limited - India', Decimal('303153'), 11),
 ('Technology Services Group', Decimal('261575'), 12),
 ('iik Technologies - Desarrollo de Soluciones de Software SA de CV', Decimal('258719'), 13),
 ('APIUX Technologia SPA.', Decimal('243048'), 14),
 ('Armedia LLC', Decimal('239108'), 15),
 ('Crest Business Solutions Pte Ltd', Decimal('196697'), 16),
 ('Amihan\xa0Global\xa0Strategies\xa0Phils\xa0Inc', Decimal('122874'), 17),
 ('Global Quark S.A. de C.V.', Decimal('113000'), 18),
 ('Contezza BV', Decimal('88715'), 19)]

In [12]:
%%sql
SELECT
    EmployeeFirstName,
    EmployeeLastName,
    SUM(ContractPrice) AS realized_sales,
    RANK() OVER (
        ORDER BY SUM(ContractPrice) DESC
    ) AS sales_rep_rank_by_sales
FROM SalesReps sr
LEFT JOIN Accounts a
    ON sr.EmployeeId = a.EmployeeId
WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31'
    AND SalesStage = 'Closed Won'
GROUP BY EmployeeFirstName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
6 rows affected.


[('Robert', 'Smith', Decimal('4974747'), 1),
 ('Katie', 'Barmen', Decimal('4524231'), 2),
 ('Rachel', 'King', Decimal('3979588'), 3),
 ('Jose', 'Gomez', Decimal('3903490'), 4),
 ('Michelle', 'Clarence', Decimal('2422192'), 5),
 ('Bobby', 'Thompson', Decimal('1816119'), 6)]

In [13]:
%%sql
CREATE OR REPLACE VIEW sales_breakdown_by_rep_region AS
SELECT
    DISTINCT sr.EmployeeId,
    EmployeeFirstName,
    EmployeeLastName,
    RegionName,
    SUM(ContractPrice) OVER (
        PARTITION BY RegionName, a.EmployeeId
    ) AS realized_sales_per_region_per_rep
FROM SalesReps sr
JOIN Accounts a
    ON sr.EmployeeId = a.EmployeeId
JOIN Countries c
    ON a.CountryId = c.CountryId
JOIN CountryRegions cr
    ON c.CountryId = cr.CountryId
JOIN Regions r
    ON cr.RegionId = r.RegionId
WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59'
    AND SalesStage = 'Closed Won'
Order BY EmployeeFirstName;

SELECT 
    DISTINCT EmployeeFirstName,
    EmployeeLastName,
    RegionName,
    realized_sales_per_region_per_rep,
    COUNT(AccountId) OVER (
        PARTITION BY RegionName, a.EmployeeId
    ) AS number_of_accounts_in_region
FROM sales_breakdown_by_rep_region sbrr
JOIN Accounts a
    ON sbrr.EmployeeId = a.EmployeeId
WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59';

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
0 rows affected.
23 rows affected.


[('Robert', 'Smith', 'AMER', Decimal('3638263'), 148),
 ('Jose', 'Gomez', 'AMER', Decimal('2668125'), 160),
 ('Rachel', 'King', 'AMER', Decimal('3515853'), 131),
 ('Bobby', 'Thompson', 'AMER', Decimal('1149983'), 127),
 ('Michelle', 'Clarence', 'AMER', Decimal('1840349'), 133),
 ('Katie', 'Barmen', 'AMER', Decimal('3222027'), 133),
 ('Robert', 'Smith', 'APAC', Decimal('556786'), 148),
 ('Jose', 'Gomez', 'APAC', Decimal('727805'), 160),
 ('Rachel', 'King', 'APAC', Decimal('367307'), 131),
 ('Bobby', 'Thompson', 'APAC', Decimal('530566'), 127),
 ('Michelle', 'Clarence', 'APAC', Decimal('204645'), 133),
 ('Katie', 'Barmen', 'APAC', Decimal('187775'), 133),
 ('Robert', 'Smith', 'EMEA', Decimal('682809'), 148),
 ('Jose', 'Gomez', 'EMEA', Decimal('507560'), 160),
 ('Rachel', 'King', 'EMEA', Decimal('32981'), 131),
 ('Bobby', 'Thompson', 'EMEA', Decimal('110408'), 127),
 ('Michelle', 'Clarence', 'EMEA', Decimal('377198'), 133),
 ('Katie', 'Barmen', 'EMEA', Decimal('760806'), 133),
 ('Robert', 'Smith', 'JPN', Decimal('96889'), 148),
 ('Katie', 'Barmen', 'JPN', Decimal('94904'), 133),
 ('Rachel', 'King', 'LATAM', Decimal('63447'), 131),
 ('Bobby', 'Thompson', 'LATAM', Decimal('25162'), 127),
 ('Katie', 'Barmen', 'LATAM', Decimal('258719'), 133)]

The first query looks at the top regions in terms of realized sales dollars for the year 2019. After query the results we find that:
* AMER is number one with 16,034,600
* APAC is number two with 2,574,884
* EMEA is number three with 2,471,762
* JPN is number four with 347,328
* LATAM is number five with 191,793

My suggestion would be to focus a lot of the attention on the region JPN and LATAM as those sales numbers for 2019 are drastically lower than EMEA, APAC, and AMER. I would start by looking at the sales strategy and marketing strategies in those regions. If the strategies check out, then it may be time to think about not selling so much those regions and just focus on increasing sales in EMEA and APAC to try and close the gap with the AMER region.

The next two queries provide a breakdown of partner performance. The first query compares how much (in terms of dollars) each of the company's partners sold in 2020 versus the company itself. The company itself is represented by None which is at the top of the list. It is good that the company is ranked number one on the list because that means that most sales are coming from in-house efforts. The second query is just a ranked list showing who the best and worst partners are for the year 2020. My only suggestion here would be for the company to potentially re-evaluate the partnership between the bottom five partners. I chose the bottom five partners because the biggest drop in terms of sales dollars occurs after Armedia LLC at 239,108. The bottom four partner companies come in at 196,697 (Crest Business Solutions Pte Ltd), 122,874 (Amihan Global Strategies Phils Inc), 113,000 (Global Quark S.A. de C.V.), and 88,715 (Contezza BV).

The fourth query just shows who the top sales performers were for the year 2020. My recommendation here would be to look at the bottom performers Bobby Thompson and Michelle Clarence and potentially put them through more sales training. I would also recommend rewarding Robert Smith accordingly as he was last year’s top performer and the company's number one revenue generator by roughly 450K+ for the year. 

The third query then breaks down each sales rep's performance in each region. Based on the generated list, all the reps except for Robert Smith and Katie Barmen were having trouble selling in Japan in 2020. In addition, Jose Gomez, Robert Smith, and Michelle Clarence did not have any LATAM sales in 2020. We also find that Tony Stamp did not sell anything in 2020. In addition, we see how many accounts each rep was working in 2020, and it seems hefty. My recommendation would be to hire more sales reps, as each rep is working 100+ accounts in almost every region putting a lot of stress on the team. The high workload may be why many sales reps did not make it to LATAM and JPN in 2020.

<center>-----------------------------------------------------------------------------------------------------------------------------------------------------</center>

##### Question 3:
 Now we know who top performers are in terms of realized sales dollar generation, but let’s compare last years closed won deals with closed lost deals. In which regions did we lose the most sales, and which sales reps are struggling to close deals? Provide a list of the customers we lost that year as well. 

##### Business Justification:
Comparing closed won and lost deals helps the company gauge the effectiveness of their sales strategy. In addition, by looking at the accounts lost it provides the sales and marketing teams an opportunity to learn from their mistakes and refine their tactics and approaches. Looking at the closed lost deals also allows the company to look at where exactly in the process they lost the potential customer, and what tactic or feature did the competitor use to pull the customer over to their product instead of ours.

##### SQL Features Used:
* CTE
* INNER & LEFT JOIN
* CASE
* GROUP BY

In [80]:
%%sql
WITH lost_rep_deals AS (
    SELECT
        EmployeeFirstName,
        EmployeeLastName,
        COUNT(AccountId) AS number_lost_accounts,
        SUM(ContractPrice) AS closed_lost_sales
    FROM Accounts a
    JOIN SalesReps sr
        ON a.EmployeeId = sr.EmployeeId
    WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59' 
        AND SalesStage = 'Closed Lost'
    GROUP BY EmployeeFirstName
    ORDER BY closed_lost_sales DESC
), 
won_rep_deals AS (
    SELECT
        EmployeeFirstName,
        EmployeeLastName,
        COUNT(AccountId) AS number_won_accounts,
        SUM(ContractPrice) AS closed_won_sales
    FROM Accounts a
    JOIN SalesReps sr
        ON a.EmployeeId = sr.EmployeeId
    WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59' 
        AND SalesStage = 'Closed Won'
    GROUP BY EmployeeFirstName
    ORDER BY closed_won_sales DESC
)
SELECT
    lrd.EmployeeFirstName,
    lrd.EmployeeLastName,
    number_lost_accounts,
    closed_lost_sales,
    number_won_accounts,
    closed_won_sales,
    CASE
        WHEN closed_lost_sales > closed_won_sales THEN '-'
        ELSE '+'
    END AS loss_vs_won_sales_status
FROM lost_rep_deals lrd
JOIN won_rep_deals wrd
    ON lrd.EmployeeFirstName = wrd.EmployeeFirstName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
6 rows affected.


[('Robert', 'Smith', 26, Decimal('5487813'), 23, Decimal('4974747'), '-'),
 ('Jose', 'Gomez', 19, Decimal('3788830'), 19, Decimal('3903490'), '+'),
 ('Bobby', 'Thompson', 16, Decimal('3522794'), 14, Decimal('1816119'), '-'),
 ('Rachel', 'King', 18, Decimal('3422597'), 20, Decimal('3979588'), '+'),
 ('Michelle', 'Clarence', 13, Decimal('2833347'), 17, Decimal('2422192'), '-'),
 ('Katie', 'Barmen', 13, Decimal('2048624'), 23, Decimal('4524231'), '+')]

In [88]:
%%sql

WITH lost_region_deals AS (
    SELECT
        RegionName,
        COUNT(AccountId) AS number_lost_accounts,
        SUM(ContractPrice) AS closed_lost_sales
    FROM Accounts a
    JOIN CountryRegions cr
        ON a.CountryId = cr.CountryId
    JOIN Regions r
        ON cr.RegionId = r.RegionId
    WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59' 
        AND SalesStage = 'Closed Lost'
    GROUP BY RegionName
    ORDER BY closed_lost_sales DESC
), 
won_region_deals AS (
    SELECT
        RegionName,
        COUNT(AccountId) AS number_won_accounts,
        SUM(ContractPrice) AS closed_won_sales
    FROM Accounts a
    JOIN CountryRegions cr
        ON a.CountryId = cr.CountryId
    JOIN Regions r
        ON cr.RegionId = r.RegionId
    WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59' 
        AND SalesStage = 'Closed Won'
    GROUP BY RegionName
    ORDER BY closed_won_sales DESC
)
SELECT
    wrd.RegionName,
    number_lost_accounts,
    closed_lost_sales,
    number_won_accounts,
    closed_won_sales,
    CASE
        WHEN closed_lost_sales > closed_won_sales THEN '-'
        ELSE '+'
    END AS loss_vs_won_sales_status
FROM won_region_deals wrd
LEFT JOIN lost_region_deals lrd
    ON lrd.RegionName = wrd.RegionName;

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
5 rows affected.


[('AMER', 75, Decimal('14116893'), 84, Decimal('16034600'), '+'),
 ('APAC', 11, Decimal('2557464'), 14, Decimal('2574884'), '+'),
 ('EMEA', 19, Decimal('4429648'), 13, Decimal('2471762'), '-'),
 ('LATAM', None, None, 3, Decimal('347328'), '+'),
 ('JPN', None, None, 2, Decimal('191793'), '+')]

In [94]:
%%sql
SELECT
    RegionName,
    AccountName,
    SalesStage
FROM Accounts a
JOIN CountryRegions cr
    ON a.CountryId = cr.CountryId
JOIN Regions r
    ON cr.RegionId = r.RegionId
WHERE ContractStartDate BETWEEN '2020-1-1' AND '2020-12-31 23:59:59'
    AND SalesStage = 'Closed Lost'

 * mysql://admin:***@lmu-dev-01.ch3fswgxeqfe.us-east-1.rds.amazonaws.com/sql_project
105 rows affected.


[('AMER', 'Calendly', 'Closed Lost'),
 ('APAC', 'Paytm', 'Closed Lost'),
 ('AMER', 'QuantumScape', 'Closed Lost'),
 ('AMER', 'Payoneer', 'Closed Lost'),
 ('EMEA', 'Camunda', 'Closed Lost'),
 ('AMER', 'Signal', 'Closed Lost'),
 ('EMEA', 'Siemens', 'Closed Lost'),
 ('EMEA', 'what3words', 'Closed Lost'),
 ('AMER', 'Axis Security', 'Closed Lost'),
 ('AMER', 'Aeva', 'Closed Lost'),
 ('APAC', 'Big Hit Entertainment', 'Closed Lost'),
 ('APAC', 'CICC', 'Closed Lost'),
 ('AMER', 'Vimeo', 'Closed Lost'),
 ('APAC', 'Info Edge', 'Closed Lost'),
 ('AMER', 'Via', 'Closed Lost'),
 ('AMER', 'Superhuman', 'Closed Lost'),
 ('AMER', 'Petco', 'Closed Lost'),
 ('EMEA', 'MPB', 'Closed Lost'),
 ('EMEA', 'Cazoo', 'Closed Lost'),
 ('AMER', 'PICKUP', 'Closed Lost'),
 ('AMER', 'Alt', 'Closed Lost'),
 ('AMER', 'Kaiser Permanente', 'Closed Lost'),
 ('AMER', 'Tetrate', 'Closed Lost'),
 ('AMER', 'Saks Fifth Avenue', 'Closed Lost'),
 ('EMEA', 'Skeleton Technologies', 'Closed Lost'),
 ('AMER', 'Riskified', 'Closed Lost'),
 ('AMER', 'PandaDoc', 'Closed Lost'),
 ('EMEA', 'Mirakl', 'Closed Lost'),
 ('EMEA', 'GetYourGuide', 'Closed Lost'),
 ('APAC', 'Ather Energy', 'Closed Lost'),
 ('AMER', 'Second Closet', 'Closed Lost'),
 ('AMER', 'Veritas Capital', 'Closed Lost'),
 ('AMER', 'CircleCI', 'Closed Lost'),
 ('AMER', 'Eden Health', 'Closed Lost'),
 ('EMEA', 'Genesis Global', 'Closed Lost'),
 ('AMER', '100Plus', 'Closed Lost'),
 ('AMER', 'HomeLight', 'Closed Lost'),
 ('APAC', 'FloBiz', 'Closed Lost'),
 ('AMER', 'GoodRx', 'Closed Lost'),
 ('EMEA', 'Cellink', 'Closed Lost'),
 ('AMER', 'Beam Therapeutics', 'Closed Lost'),
 ('AMER', 'Endless West', 'Closed Lost'),
 ('AMER', 'FreightWaves', 'Closed Lost'),
 ('EMEA', 'Decibel', 'Closed Lost'),
 ('EMEA', 'Debenhams', 'Closed Lost'),
 ('AMER', 'Mason', 'Closed Lost'),
 ('AMER', 'CareDx', 'Closed Lost'),
 ('AMER', 'Heal', 'Closed Lost'),
 ('AMER', 'Jetty', 'Closed Lost'),
 ('AMER', 'Biz2Credit', 'Closed Lost'),
 ('AMER', 'Simple Health', 'Closed Lost'),
 ('AMER', 'Click Therapeutics', 'Closed Lost'),
 ('AMER', 'Swyft', 'Closed Lost'),
 ('EMEA', 'Abzu', 'Closed Lost'),
 ('AMER', 'GrowGeneration', 'Closed Lost'),
 ('AMER', 'AgBiome', 'Closed Lost'),
 ('AMER', 'FreshDirect', 'Closed Lost'),
 ('AMER', 'SparkPost', 'Closed Lost'),
 ('AMER', 'Fivestars', 'Closed Lost'),
 ('APAC', 'Nreal', 'Closed Lost'),
 ('AMER', 'Stocktwits', 'Closed Lost'),
 ('APAC', 'Kyber Network', 'Closed Lost'),
 ('AMER', 'Clever', 'Closed Lost'),
 ('AMER', 'JUUL', 'Closed Lost'),
 ('AMER', 'QVC', 'Closed Lost'),
 ('AMER', 'PicsArt', 'Closed Lost'),
 ('AMER', 'Capchase', 'Closed Lost'),
 ('AMER', 'Listo', 'Closed Lost'),
 ('AMER', 'Mammoth Biosciences', 'Closed Lost'),
 ('AMER', 'Ladder', 'Closed Lost'),
 ('AMER', 'HoneyBook', 'Closed Lost'),
 ('AMER', 'Rebag', 'Closed Lost'),
 ('AMER', 'OpenWeb', 'Closed Lost'),
 ('APAC', 'Trell', 'Closed Lost'),
 ('AMER', 'Woflow', 'Closed Lost'),
 ('AMER', 'TrustRadius', 'Closed Lost'),
 ('EMEA', 'TradingView', 'Closed Lost'),
 ('EMEA', 'Maze', 'Closed Lost'),
 ('AMER', 'Digbi Health', 'Closed Lost'),
 ('AMER', 'DrChrono', 'Closed Lost'),
 ('AMER', 'Enigma Technologies', 'Closed Lost'),
 ('AMER', 'Token Transit', 'Closed Lost'),
 ('AMER', 'Earnest', 'Closed Lost'),
 ('EMEA', 'Splyt', 'Closed Lost'),
 ('AMER', 'LambdaTest', 'Closed Lost'),
 ('AMER', 'Zeus Living', 'Closed Lost'),
 ('AMER', 'Framework Ventures', 'Closed Lost'),
 ('AMER', 'Descartes Labs', 'Closed Lost'),
 ('AMER', 'RumbleON', 'Closed Lost'),
 ('AMER', 'Embark', 'Closed Lost'),
 ('EMEA', 'Signal AI', 'Closed Lost'),
 ('AMER', 'Boston Dynamics', 'Closed Lost'),
 ('AMER', 'Nature’s Fynd', 'Closed Lost'),
 ('AMER', 'Fauna', 'Closed Lost'),
 ('AMER', 'FreeWheel', 'Closed Lost'),
 ('AMER', 'L7 Informatics', 'Closed Lost'),
 ('APAC', 'GetWork', 'Closed Lost'),
 ('EMEA', 'StorageOS', 'Closed Lost'),
 ('AMER', 'TriNet', 'Closed Lost'),
 ('AMER', 'Quicklly', 'Closed Lost'),
 ('AMER', 'Simplilearn', 'Closed Lost'),
 ('AMER', 'Kenna Security', 'Closed Lost'),
 ('EMEA',

In a nutshell the first two queries outline the number of accounts lost/won and the dollar amount of lost/won accounts both by sales rep and region. In addition, for both queries I added a CASE statement to indicate whether the dollar amount of lost sales is larger than the dollar amount of won sales for the year 2020. a '+' means that the dollar amount of closed won deals is greater than the dollar amount of closed lost deals and the '-' means the opposite.

We see that Katie Barmen, Rachel King, and Jose Gomez all have realized sales for 2020 greater than the dollar amount they lost. On the other hand, Robert Smith, Bobby Thompson, and Michelle Smith all lost more sales dollars than they gained. In addition, we also see that the only region that was not profitable for the company in 2020 was EMEA.  

The third query is just a list of customers who decided not to buy the company's product.

Ultimately my recommendation would be for the sales executives to debrief with all their reps starting with the ones that closed the year at a loss. Have them discuss what is and is not working in terms of sales tactics and/or product pricing and positioning. I would also suggest that the company re-evaluate the EMEA market to see why the region was not profitable in 2020. Also, if possible I would also recommend that the sales team look back at the phone recordings for the lost customers and review the notes in SFDC to try and pinpoint exactly where they lost the customer in the process.